core/preconditioner/isai.cpp

```diff
@@ -265,58 +266,84 @@ void Isai<IsaiType, ValueType, IndexType>::generate_inverse(
 }
 
 
-template <isai_type IsaiType, typename ValueType, typename IndexType>
-std::unique_ptr<LinOp> Isai<IsaiType, ValueType, IndexType>::transpose() const
+template <isai_type IsaiType, typename ValueType, typename IndexType,
+          typename StorageType>
+std::unique_ptr<LinOp>
+Isai<IsaiType, ValueType, IndexType, StorageType>::transpose() const
 {
+    auto exec = this->get_executor();
     auto is_spd = IsaiType == isai_type::spd;
     if (is_spd) {
         return this->clone();
     }
 
-    std::unique_ptr<transposed_type> transp{
-        new transposed_type{this->get_executor()}};
+    std::unique_ptr<transposed_type> transp{new transposed_type{exec}};
     transp->set_size(gko::transpose(this->get_size()));
-    transp->approximate_inverse_ =
-        share(as<Csr>(this->get_approximate_inverse())->transpose());
+    auto csr_transp =
+        share(as<Csr>(as<Csr>(this->get_approximate_inverse())->transpose()));
```

Comment:
In general in this PR: you use too much `as<>()` and `share()`, which makes it harder to read.